File: medium/1405. Longest Happy String.py
# 1405. Longest Happy String
class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        '''
        Buckets.
        Similar to 358. Rearrange String k Distance Apart
        and 984. String Without AAA or BBB.
        '''
        (first, x), (second, y), (third, z) = sorted(
            [(a, 'a'), (b, 'b'), (c, 'c')], reverse=True)
        buckets = [x + x] * (first // 2)
        if first % 2 == 1:
            buckets.append(x)
        num_buckets = len(buckets)
        bucket_idx = 0
        for count, char in [(second, y), (third, z)]:
            for _ in range(count):
                buckets[bucket_idx % num_buckets] += char
                bucket_idx += 1
        # There may be no enough separator between each bucket, so we only
        # take the ones that are separated.
        return ''.join(buckets[:min(num_buckets, bucket_idx + 1)])

File: medium/1405. Longest Happy String.rb
# 1405. Longest Happy String
# @param {Integer} a
# @param {Integer} b
# @param {Integer} c
# @return {String}
def longest_diverse_string(a, b, c)
    (first, x), (second, y), (third, z) = [[a, 'a'], [b, 'b'], [c, 'c']].sort.reverse
    buckets = [x + x] * (first / 2)
    buckets.push(x) if first % 2 == 1
    num_buckets = buckets.length
    bucket_idx = 0
    [[second, y], [third, z]].each do |count, char|
        (0...count).each do |_|
            buckets[bucket_idx % num_buckets] += char
            bucket_idx += 1
        end
    end

    buckets[...[num_buckets, bucket_idx+1].min].join
end